[RFC] Don't materialize ignored modules for FSDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108435
* #108033
* __->__ #108032

Per title. This seems needed for cases where I have a large embedding
I want to separately manage, but FSDP would initialize it and thus consume the
memory.

Currently the interaction with torchdistX materialize_module is not tested,
this can be done as follow up work.

Differential Revision: [D48722046](https://our.internmc.facebook.com/intern/diff/D48722046/)

rohan-varma committed:
[RFC] Don't materialize ignored modules for FSDP

Per title. This seems needed for cases where I have a large embedding
I want to separately manage, but FSDP would initialize it and thus consume the
memory.

Currently the interaction with torchdistX materialize_module is not tested,
this can be done as follow up work.

Differential Revision: [D48722046](https://our.internmc.facebook.com/intern/diff/D48722046/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -640,6 +640,33 @@ def __init__(self):
+            device_id=self.rank,
+            ignored_modules=[m.a],
+            use_orig_params=True,
+            param_init_fn=lambda m: m.cuda(),
```
In general, this `lambda m: m.cuda()` would do some repeated checks trying to move modules to CUDA since `param_init_fn` would be called on every module. This should just lead to some CPU overhead since copying to CUDA is a no-op if already on CUDA.

As a tiny nit, the variable shadowing of `m` is also a bit precarious.

awgu reviewed (approved):
SGTM!

awgu reviewed (changes requested):
Sorry, I should cancel the approve. The unit test is failing.

```
return self._apply(lambda t: t.cuda(device))
NotImplementedError: Cannot copy out of meta tensor; no data!
```
I think you need a different `param_init_fn`.

rohan-varma commented:
That's weird, I feel like I ran the test before sending the PR...

rohan-varma committed:
Update on "[RFC] Don't materialize ignored modules for FSDP"


Per title. This seems needed for cases where I have a large embedding
I want to separately manage, but FSDP would initialize it and thus consume the
memory.

Currently the interaction with torchdistX materialize_module is not tested,
this can be done as follow up work.

Differential Revision: [D48722046](https://our.internmc.facebook.com/intern/diff/D48722046/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Don't materialize ignored modules for FSDP"


Per title. This seems needed for cases where I have a large embedding
I want to separately manage, but FSDP would initialize it and thus consume the
memory.

Currently the interaction with torchdistX materialize_module is not tested,
this can be done as follow up work.

Differential Revision: [D48722046](https://our.internmc.facebook.com/intern/diff/D48722046/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Don't materialize ignored modules for FSDP"


Per title. This seems needed for cases where I have a large embedding
I want to separately manage, but FSDP would initialize it and thus consume the
memory.

Currently the interaction with torchdistX materialize_module is not tested,
this can be done as follow up work.

Differential Revision: [D48722046](https://our.internmc.facebook.com/intern/diff/D48722046/)

[ghstack-poisoned]
